lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -128,30 +150,43 @@ class AllocaPattern : public OpRewritePattern<cudaq::cc::AllocaOp> {
       for (auto &useuse : user->getUses()) {
         auto *useuser = useuse.getOwner();
         if (auto ist = dyn_cast<quake::InitializeStateOp>(useuser)) {
+          rewriter.setInsertionPointAfter(useuser);
           LLVM_DEBUG(llvm::dbgs() << "replaced init_state\n");
           assert(conGlobal && "global must be defined");
           rewriter.replaceOpWithNewOp<quake::InitializeStateOp>(
               ist, ist.getType(), ist.getTargets(), conGlobal);
           continue;
         }
         if (auto load = dyn_cast<cudaq::cc::LoadOp>(useuser)) {
+          rewriter.setInsertionPointAfter(useuser);
           LLVM_DEBUG(llvm::dbgs() << "replaced load\n");
           rewriter.replaceOpWithNewOp<cudaq::cc::ExtractValueOp>(
               load, eleTy, conArr,
               ArrayRef<cudaq::cc::ExtractValueArg>{offset});
           continue;
         }
         if (isa<cudaq::cc::StoreOp>(useuser))
-          rewriter.eraseOp(useuser);
+          toErase.push_back(useuser);
         isLive = true;
       }
       if (!isLive)
-        rewriter.eraseOp(user);
+        toErase.push_back(user);
     }
     if (toGlobal) {
+      rewriter.setInsertionPointAfter(alloc);
       rewriter.replaceOp(alloc, conGlobal);
     } else {
-      rewriter.eraseOp(alloc);
+      toErase.push_back(alloc);
+    }
+
+    for (auto *op : toErase) {
+      if (op->getUses().empty()) {
```

Comment:
Stores will never have uses.

Reply:
I think we are removing `alloc` sometimes, that's where I saw it fail once. Just wanted an error message to find if it happens again, but I can remove this if you prefer.

Reply:
removed